### PR TITLE
ref(kafka): Make batching_kafka_consumer work with existing consumer

### PR DESCRIPTION
**Pull request by nikhars**
We want the ability to have BatchingKafkaConsumer work with a consumer
which is created outside of it. The reason being that the current
BatchingKafkaConsumer only provides a mechanism to create and use one.

The change includes creating a Facade with the API's necessary to
work with the BatchingKafkaConsumer and allowing it to use an exisitng
consumer if it exists. Also changed SynchronizedConsumer to be an
implementation of the Facade

### CODE REVIEW

File: src/sentry/utils/batching_kafka_consumer.py [hunk 1]
```diff
@@ -177,18 +216,33 @@ def __init__(
...
+            self.consumer = self.create_consumer(
+                topics,
+                cluster_name,
+                group_id,
+                auto_offset_reset,
+                queued_max_messages_kbytes,
+                queued_min_messages,
+            )
+        else:
+            assert isinstance(consumer, KafkaConsumerFacade)
```
Comment (fpacifici) [inline src/sentry/utils/batching_kafka_consumer.py:232]: Please set the type hint for the consumer parameter, so you do not need the isinstance.

File: src/sentry/utils/batching_kafka_consumer.py [hunk 2]
```diff
@@ -177,18 +216,33 @@ def __init__(
...
+            self.consumer = consumer
+
+        def on_partitions_assigned(consumer, partitions):
+            logger.info("New partitions assigned: %r", partitions)
```
Comment (fpacifici) [inline src/sentry/utils/batching_kafka_consumer.py:236]: Isn't the post process forwarder going to need to provide its own `on_partition_assigned` and `on_partition_revoked` methods ?

**Review by fpacifici — APPROVED** [review]
Few minor comments. 
Specifically the one on the assign/revoke callback does not impact this PR but it will impact you when you use this in the post process forwarder.

File: src/sentry/utils/batching_kafka_consumer.py (hunk 1 above)
Comment (nikhars) [reply src/sentry/utils/batching_kafka_consumer.py:232]: This file is not in mypy.ini. That's why I added the runtime check here

File: src/sentry/utils/batching_kafka_consumer.py (hunk 2 above)
Comment (nikhars) [reply src/sentry/utils/batching_kafka_consumer.py:236]: I am not sure if its needed. I will handle this case during the integration PR if needed. Some context. 

The batching kafka consumer calls
self.consumer.subscribe at https://github.com/getsentry/sentry/pull/28677/files#diff-0baa7d2f6a7029fbcf066aacd1590491a129a3e20a6e7444005aca32b0ad0778R243-R245

Which in turn calls https://github.com/getsentry/sentry/blob/master/src/sentry/eventstream/kafka/consumer.py#L266-L325 This has its own assignment and revocation methods. Those methods call the additional callbacks provided as arguments above.

The flush method in the revocation callback at https://github.com/getsentry/sentry/pull/28677/files#diff-0baa7d2f6a7029fbcf066aacd1590491a129a3e20a6e7444005aca32b0ad0778R241 calls self._commit at https://github.com/getsentry/sentry/blob/master/src/sentry/utils/batching_kafka_consumer.py#L387 which calls the consumer commit method here https://github.com/getsentry/sentry/blob/master/src/sentry/utils/batching_kafka_consumer.py#L401 which takes care of committing the offsets.

I will do some additional testing though during the integration though to validate things

**Pull request merged**